### PR TITLE
Eigenstates sparse orthogonalization.

### LINKED ISSUE #1851: Orthogonalization step in `eigs_csr` missing
### Bug Description

In `dev.major`, `eigs_csr` produce eigenvectors that are not orthogonal when sharing the same eigenvalue.

### Code to Reproduce the Bug

```shell
import qutip
import numpy as np
H = qutip.rand_herm([1, 1, 1, 2, 2, 2])
_, evecs = qutip.data.eigs(H.to('csr').data, 1, 1)
evecs = evecs.to_array()

for i in range(6):
    vec1 = evecs[:,i]
    for j in range(i+1, 6):
        vec2 = evecs[:,j]
        overlap = np.inner(vec1, vec2.conj())
        if not abs(overlap) < 1e-7:
            print(i, j, overlap)
```

### Code Output

```shell
0 1 (0.09845037882574664-0.5162235950159674j)
0 2 (0.26622727906522153+0.05779924588623911j)
1 2 (-0.03215914426418319+0.03471465108382346j)
3 4 (0.12577654493163992+0.29446679788486496j)
3 5 (0.06208346730638102-0.24881812976497342j)
4 5 (0.02714640795325415+0.20205107594878488j)
```

### Expected Behaviour

While the documentation does not says the eigenvectors would be orthogonal and normalized, it is expected and more practical this way.

### Your Environment

```shell
QuTiP Version:      5.0.0.dev0+0619b3b
Numpy Version:      1.22.2
Scipy Version:      1.7.3
Python Version:     3.9.7
BLAS Info:          OPENBLAS
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
 For hermician matrices, eigenstates are expected to be orthogonal to each other. 

`eigh` used by the dense specialization create such states.  But in `eigs_csr`, since we obtain the states in 2 calls of `sp.linalg.eigsh`, the states can overlap.

This ensure that the eigen states produced by `eigs_csr` are orthogonal and add a test for it.

**Related issues or PRs**
Fix #1851

### CODE REVIEW

File: qutip/core/data/eigen.py [hunk 1]
```diff
@@ -225,6 +227,15 @@ def eigs_csr(data, isherm=None, vecs=True, sort='low', eigvals=0,
     isherm = isherm if isherm is not None else _isherm(data)
     evals, evecs = _eigs_csr(data.as_scipy(), isherm, vecs, eigvals,
                              num_large, num_small, tol, maxiter)
+
+    if vecs and isherm:
+        i = 0
+        while i < len(evals):
+            num_degen = np.sum(np.abs(evals - evals[i]) < (2 * tol or 1e-14))
+            for k, l in combinations(range(num_degen), 2):
+                _orthogonalize(evecs[:, i+l], evecs[:, i+k])
```
Comment (hodgestar) [inline qutip/core/data/eigen.py:236]:
```suggestion
            # orthogonalize vectors 1 .. k with respect to the first, then 2 .. k with respect to the second, and so on.
            # relies on both the order of each pair and the ordering of pairs returned by combinations. 
            for k, l in combinations(range(num_degen), 2):
                _orthogonalize(evecs[:, i+l], evecs[:, i+k])
```
The ordering requirements used here are nice, but also non-obvious, so I suggested a comment.

**Review by hodgestar — CHANGES_REQUESTED** [review]
The fix looks good, but the original tests seemed a bit odd and the new test inherited some of that oddness.

File: qutip/tests/core/test_eigenstates.py [hunk 2]
```diff
@@ -68,15 +69,21 @@ def test_known_eigensystem(hamiltonian, eigenvalues, eigenstates):
 
 # Specify parametrisation over a random Hamiltonian by specifying the
 # dimensions, rather than duplicating that logic.
-@pytest.fixture(params=[pytest.param([5], id="simple"),
+@pytest.fixture(params=[pytest.param([10], id="simple"),
                         pytest.param([5, 3, 4], id="tensor")])
 def random_hamiltonian(request):
     dimensions = request.param
-    return qutip.tensor(*[qutip.rand_herm(dim) for dim in dimensions])
+    return qutip.tensor(*[qutip.rand_herm(dim, dtype='csr')
```
[the remark this answers is not in the document]
Comment (hodgestar) [reply qutip/tests/core/test_eigenstates.py]: By "fix" I meant "not use the word dim and instead use the word evals". :)

[the remark this answers is not in the document]
Comment (hodgestar) [reply qutip/tests/core/test_eigenstates.py]: Hmm. With the new change, I don't know if there is a single word that describes what the parameter is. It's either "dimensions" or "lists_of_evecs" depending on whether it is a list of integers or a list of lists. :/

File: qutip/tests/core/test_eigenstates.py [hunk 3]
```diff
@@ -68,15 +69,24 @@ def test_known_eigensystem(hamiltonian, eigenvalues, eigenstates):
...
-@pytest.fixture(params=[pytest.param([5], id="simple"),
-                        pytest.param([5, 3, 4], id="tensor")])
+@pytest.fixture(params=[pytest.param([10], id="simple"),
+                        pytest.param([5, 3, 4], id="tensor"),
+                        pytest.param([[0, 1, 1]]*3, id="degenerate")])
 def random_hamiltonian(request):
     dimensions = request.param
-    return qutip.tensor(*[qutip.rand_herm(dim) for dim in dimensions])
+    return qutip.tensor(*[qutip.rand_herm(dim)
+                          for dim in dimensions])
```
Comment (Ericgig) [inline qutip/tests/core/test_eigenstates.py:78]:
```suggestion
    return qutip.tensor(*[qutip.rand_herm(dim_or_evecs)
                          for dim_or_evecs in dimensions])
```

**Review by hodgestar — APPROVED** [review]

**Pull request merged**